### PR TITLE
Allow disabling of formatting while typing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This is a potential workaround for https://github.com/dotnet/roslyn/issues/8224 and https://github.com/dotnet/roslyn/issues/5873

Provided here is a way for users to completely opt out of the C# language service doing any formatting that wasn't explicitly asked for.  The way to do that is now to disable all of these options:

image

image

Note: with all those options off, the C# language service will still automatically pick an appropriate indentation amount when you hit enter.  If even that behavior is undesirable it can be turned off with this option:

image

image

Picking 'None' or 'Block' means that C# will not even try to determine the appropriate indentation level for your code.

With all these options off, you should be in total control of your code.

**Comment by CyrusNajmabadi** [comment]
This is based on the requests of several people (like @atykhyy) who mentioned

> Yes, please! Auto-formatting works fine for many people, but there is a large minority for whom it's incredibly frustrating. A simple option to turn off formatting completely would be a godsend.

And @craigs100 

> Either very clear instructions are needed as to how to set up auto-formatting, including pastes and the options actual effects described, or a facility to switch it ALL off.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit64_prtest  please

**Comment by CyrusNajmabadi** [comment]
Tagging @mavasani and @heejaechang as well.  This is the PR that motivated the email i sent you about complexity between the AbstractIndentationService, the SmartTokenFormatterCommandHAndlerService, and the FormatCommandHandlerService.  It's hard to rationalize what they're all responsible for.  And, as such, it's difficult to know if this change actually properly disables formatting when the user wants it.

I'm hoping we can really refactoring things in this area to make it clear that indentation is only handled by one service (and is configurable through one set of options), and formatting is only handled by one service (and is totally configurable through other independent options).

File: src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs [hunk 1]
```diff
@@ -35,21 +36,42 @@ public CSharpEditorFormattingService()
...
         {
             var options = document.GetOptionsAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None);
             var smartIndentOn = options.GetOption(FormattingOptions.SmartIndent) == FormattingOptions.IndentStyle.Smart;
 
-            if ((ch == '}' && !options.GetOption(FeatureOnOffOptions.AutoFormattingOnCloseBrace) && !smartIndentOn) ||
-                (ch == ';' && !options.GetOption(FeatureOnOffOptions.AutoFormattingOnSemicolon)))
+            // We consider the proper placement of a close curly when it is typed at the start of the
+            // line to be a smart-indentation operation.  As such, even if "format on typing" is off,
+            // if "smart indent" is on, we'll still format this.  (However, we won't touch anything
+            // else in teh block this close curly belongs to.).
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs]: You clearly hate the word 'the'

File: src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs [hunk 2]
```diff
@@ -35,21 +36,42 @@ public CSharpEditorFormattingService()
...
+            //
+            // TODO(cyrusn): Should we expose an option for this?  Personally, i don't think so.
+            // If a user doesn't want this behavior, they can turn off 'smart indent' and control
+            // everything themselves.
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs]: I agree with your opinion in the comment above.

File: src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs [hunk 3]
```diff
@@ -173,6 +201,8 @@ private static bool TokenShouldNotFormatOnReturn(SyntaxToken token)
 
         public async Task<IList<TextChange>> GetFormattingChangesAsync(Document document, char typedChar, int caretPosition, CancellationToken cancellationToken)
         {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs]: Why move this here if it's not used until much later? There's a lot of checks that might bail out before we would even need the options.

File: src/EditorFeatures/CSharpTest/Formatting/FormattingEngineTests.cs [hunk 4]
```diff
@@ -1166,6 +1166,160 @@ class C
...
+    class C
+    {
+        int x   =   10     ;
+    }
+}
+";
+
+            var optionSet = new Dictionary<OptionKey, object>
+            {
+                    { new OptionKey(FeatureOnOffOptions.AutoFormattingOnTyping, LanguageNames.CSharp), false }
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Formatting/FormattingEngineTests.cs]: Nit: additional indent

File: src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs]: indeed...

**Comment by CyrusNajmabadi** [comment]
test this please

**Comment by CyrusNajmabadi** [comment]
retest windows_release_unit64_prtest  please

**Comment by craigs100** [comment]
The ideal is to be able to set your own indent, not have it done for you. i.e. all I really want is:-
1) Indent a line with tabs, write some code, hit return and have the cursor appear at the same tab position on the next line.
2) If I open a curly brace and hit return, I don't want the curly brace moving. I put it there because that's where I want it. Please leave it alone.
3) If I paste a code block or anything else from a text file, somewhere else on the same page or anywhere, I don't want it, or anything else on the page changing. I'll organise my newly pasted code how I want.
I want the above to happen in all editors I use and for all languages. Basically we just just a TEXT editor, code is nothing special. And I'd rather not have to have VS need to use the nuclear option of graphics accelerators just put put up a text editor. KISS principle applies. Slows the whole show down in a VM.

None of this has improved with VS2017 which was VERY dissappointing :(

Many thanks for listening :)

**Comment by Pilchie** [comment]
@craigs100 Did you try changing your settings as @CyrusNajmabadi mentioned above?

**Comment by craigs100** [comment]
@Pilchie Yes. Just to double check, having checked the settings and restarted VS, I added an else statement to an if statement and had everything suddenly hard left align. This was in a cshtml view file. So no change. I have the impression these formatters are confounded when using mixed code pages such as razor views where you can have html, js, css and c# coexisting.

**Comment by CyrusNajmabadi** [comment]
@craigs100 Can you give me a step by step set of repro steps for what you are seeing, and what you *expect* to see?  A video might also be helpful.  When it comes to formatting sometimes it's not clear what behavior people are referring to.

Thanks!

**Comment by CyrusNajmabadi** [comment]
> Indent a line with tabs, write some code, hit return and have the cursor appear at the same tab position on the next line.

This should be what the "Block" indentation setting gets you.  Effectively, indentation will always be 
identical to the indentation of the previous line.  If you're not seeing this, please give a simple repro so we can take a look!

> If I open a curly brace and hit return, I don't want the curly brace moving. I put it there because that's where I want it. Please leave it alone.

Please provide a repro for when you see this.  Please also provide your current formatting/indentation settings.  Thanks!

> None of this has improved with VS2017 which was VERY dissappointing 

Things will change if we can get repros of the behavior you're seeing so that we can fix things, and validate the scenario works properly for you. Otherwise, the behavior is just going to stay the same :)

**Comment by craigs100** [comment]
@CyrusNajmabadi Thanks for the response. Setting the Tab controls to "Block" and "On" has sorted out one issue for me so that now goes away. Although the whole UI around formatting is very confusing and could do with some serious decluttering of dependencies and duplication.
I will endeavour to provide a video of the other formatting behaviours which are still present.

**Pull request merged**